### PR TITLE
feat: filetype.lua

### PR DESCRIPTION
**Pull request by gpanders**
This is a first (rough) draft of what a `filetype.lua` might look like. Some notes:

1. ~~This (currently) only supports filetypes based off of file extension~~
2. The`vim.cmd` is ugly, but is only necessary until Lua autocommands are supported
3. ~~I don't like the `vim.tbl_extend`, but it's necessary because the `filetype.lua` in the Neovim runtime runs *after* `filetype.lua` in the user's runtime. The user may have already created a `vim.filetype` table, and we don't want to clobber it. Maybe there is a better way to do this.~~
4. ~~There is no way to define priority with this method. The existing method (autocommands) can create an ordering by modifying the order which autocommands are defined, but we can't do that here. This implementation always defaults to using the file extension first and only falls back to the other methods if file extension doesn't work. This is best for performance, but could (maybe?) have some undesirable side effects (we'll see).~~

Related: #16512

This is intended as an *extension* of standard filetype detection from `filetype.vim`, **not** a replacement. For example, with this PR users can create their own filetype detection by creating `~/.config/nvim/filetype.lua` with the contents (example):

**EDIT: Note the examples in this comment may be out-of-date, be sure to read through the rest of the discussion and the actual diff before making any comments on the API**

```lua
vim.filetype.add({
  extension = {
    foo = "fooscript",
  },
})
```

The value in the filetype table can also be a function:

```lua
vim.filetype.add({
  extension = {
    h = function(path, bufnr)
      local first_line = vim.api.nvim_buf_get_lines(bufnr, 0, 1, true);
      if first_line:match('<iostream>') then
        return "cpp"
      end
      return "c"
    end,
  }
})
```

More examples [here](https://github.com/neovim/neovim/pull/16600#issuecomment-990409210).

`filetype.lua` is sourced before `filetype.vim` so any filetypes defined in `filetype.lua` will take precedence.

### CODE REVIEW

**Comment by clason** [comment]
I think the convention has been so far "vim first, then lua" when both files were sourced from a location?

**Comment by gpanders** [comment]
> I think the convention has been so far "vim first, then lua" when both files were sourced from a location?

This is true, but if we want the catch-all autocommand to fire before the ones in `filetype.vim`, we need to source `filetype.lua` first.

**Comment by clason** [comment]
To be perfectly honest, I never fully understood the rationale behind this choice -- I would have picked the other way around, too.

**Comment by gpanders** [comment]
> But there's a difference between the "core" runtime files and the user's runtime path -- would it be possible for the core to come first while still loading user's filetype.vim before filetype.lua?

I don't think so, since it just looks it up using the runtimepath mechanism, which is always going to use the user's runtime before the core one.

>(In particular, we probably don't want a filetype.vim in the default runtime anymore?)

Down the road perhaps, but adding this as an extension mechanism (a way to define *new* filetype detection) first means we can ship it sooner and we can work out the kinks/ergonomics before going in full bore.

**Comment by clason** [comment]
That all makes sense. Maybe add a comment to the code about the rationale for the divergence?

**Comment by bfredl** [comment]
> I don't think so, since it just looks it up using the runtimepath mechanism, which is always going to use the user's runtime before the core one.

If this is a concern, we should replace

    source_runtime(FILETYPE_FILE_LUA, DIP_ALL);
    source_runtime(FILETYPE_FILE_VIM, DIP_ALL);

with a pattern that sources _both_ `filetype.lua` and `filetype.vim` for each rtp entry in turn (similar to how `mod.lua` and `mod/init.lua` are considered for first rtp entry, and then the next, and so forth).

**Comment by gpanders** [comment]
That's good point @bfredl, as this stands now the current source order is

$USER/filetype.lua
$VIMRUNTIME/filetype.lua
$USER/filetype.vim
$VIMRUNTIME/filetype.vim

when it should (probably) be

$USER/filetype.lua
$USER/filetype.vim
$VIMRUNTIME/filetype.lua
$VIMRUNTIME/filetype.vim

**Comment by clason** [comment]
The default filetype scripts have two more possible resolvers:
1. full name, like `README`;
2. complex path globs, like `.git/config`.

Do we want to support those as well as first-class citizens, or just say "use a function" (possibly offering some convenience functions for standard patterns)?

**Comment by gpanders** [comment]
> The default filetype scripts have two more possible resolvers:
> 
>     1. full name, like `README`;
> 
>     2. complex path regex, like `.git/config`.
> 
> 
> Do we want to support those as well as first-class citizens, or just say "use a function" (possibly offering some convenience functions for standard patterns)?

We already pass the full file path as the first argument of the function, so I want to say that (2) is already possible. We should support files without an extension as well. I'll think about what that API might look like.

**Comment by gpanders** [comment]
I added filename (and pattern matching) in the latest commit.

Example:

```lua
vim.filetype.add({
    filename = {
        ["foo.(%a+)"] = function(path, bufnr, ext)
            return ext
        end,
    },
})
```

Opening `foo.xyz` sets filetype `xyz`; opening `foo.awesome` sets filetype to `awesome`. This is a contrived example, but hopefully it demonstrates the power of the interface.

More realistic examples:

```lua
vim.filetype.add({
    filename = {
        [".git/config"] = "gitconfig",
        ["~/.config/mutt/muttrc"] = "muttrc",
        ["README$"] = function(path, bufnr)
            if string.find("#", vim.api.nvim_buf_get_lines(bufnr, 0, 1, true)) then
                return "markdown"
            end

            -- no return means the filetype won't be set and to try the next method
        end,
    },
})
```

**Comment by muniter** [comment]
This looks great!.

Considering that one of the motivations for this is to make things more performant we should consider some more strategies that at the same time makes things not too complex.

Providing the ability to use functions seems great, but a 100 function checks that are going to make one API call each to retrieve the same information first, second, last line, etc., could be avoided. If we go by the more generic route of providing a table to the function and we fill this table previously with the usual information a filetype needs.

For example providing a table with the following fields:

- extension
- path
- path tail
- path pieces (maybe)
- first 4 lines (or less)
- last 4 lines (or less)

Now filetypes only need a simple table lookup to get the information they need. Has the added benefit of being able to add even more things later as new pattern might emerge, the table can also be lazy loaded, or just filled when starting to match filetypes that provide a function, etc.

A disadvantage of this approach is the mutability of the table, but we could think of making a read-only table for this use case.

Just my thoughts on taking advantage of the redesign to make it a design that scales really well.

**Comment by gpanders** [comment]
>Providing the ability to use functions seems great, but a 100 function checks that are going to make one API call each to retrieve the same information first, second, last line, etc., could be avoided

Reading buffer contents is pretty rare in filetype detection, so I don't think this is an issue in practice.

The vast majority of detection happens by file detection. Another huge chunk happens based on path/filename. Only a small minority actually read the buffer or do something more complex.

I just did a quick manual count of `filetype.vim` and found fewer than 40 autocommands that call `getline()` out of 736 currently defined in that file.

**Comment by gpanders** [comment]
The `gen_filetype.lua` script autogenerates all of the (simple) extension mappings for us. Extensions that call a function (e.g. `c`) are not included.

I also (manually) ported some of the filename based matches. I did not do all of them: there are quite a bit. But the good thing is that we don't *need* all of them, since we still fall back to `filetype.vim` for now. So we can port those over time.

I did not intend for this PR to be a replacement of `filetype.vim`, so maybe we can back the `filename` matches out of this PR. Open to hear what others think.

**Comment by clason** [comment]
I think we should not greedily extract information, but at the same time having to compute the same multiple times is also bad.

One compromise that the separate table approach allows is to do them in sequence, and compute and store information at the beginning:

1. Check literal filename (only needs the filename)
2. Check extension (only needs the extension)
3. Check path (only needs the full path, possibly segmented)
4. Check contents (needs contents; could use a cache that is grown, so each time a check asks for more lines, add them to the cache)?

For that reason, I'm not 100% happy with lumping the literal filename and the more complex regex into the same category.

(And -- but this should be benchmarked -- the biggest improvement is reducing the creation of 800 autocommand to just one, followed by making the extension-based check a simple lookup rather than a regex match. So it's maybe not as critical to optimize the remaining parts at any cost?)

**Comment by gpanders** [comment]
>For that reason, I'm not 100% happy with lumping the literal filename and the more complex regex into the same category.

Yes I agree: I separated plain filenames from Lua pattern matching, so `vim.filetype.add()` now takes 3 subtables: `extension`, `filename`, and `pattern`.

I also updated the matching order to first match against full file path, then just the file name, then the extension, and finally pattern matching after that.

**Comment by clason** [comment]
Note: Needs a way to (manually) add the "complex" filetype autocommands from `filetype.vim` without them being clobbered by the script. Could put them in a different table and concatenate them?

Alternatively, it may be easier to always manually add the new filetype detections from upstream -- it's only adding a new key to the table, after all...

File: src/nvim/lua/vim.lua [hunk 1]
```diff
@@ -663,4 +663,6 @@ vim._expand_pat_get_parts = function(lua_string)
   return parts, search_index
 end
 
+vim.filetype = require('vim.filetype')
```
Comment (clason) [inline src/nvim/lua/vim.lua:666]: Are you sure it works like that? CI _really_ doesn't like this line ;)

File: runtime/lua/vim/filetype.lua [hunk 2]
```diff
@@ -0,0 +1,1111 @@
...
+  ["/etc/profile"] = function() vim.fn["dist#ft#SetFileTypeSH"](vim.fn.getline(1)) end,
+  APKBUILD = function() vim.fn["dist#ft#SetFileTypeSH"]("bash") end,
+  PKGBUILD = function() vim.fn["dist#ft#SetFileTypeSH"]("bash") end,
+  ["bash.bashrc"] = function() vim.fn["dist#ft#SetFileTypeSH"]("bash") end,
+  bashrc = function() vim.fn["dist#ft#SetFileTypeSH"]("bash") end,
+  crontab = function() vim.fn["s:StarSetf"]('crontab') end,
+  ["csh.cshrc"] = vim.fn["dist#ft#CSH"],
+  ["csh.login"] = vim.fn["dist#ft#CSH"],
+  ["csh.logout"] = vim.fn["dist#ft#CSH"],
+  ["indent.pro"] = function() vim.fn["dist#ft#ProtoCheck"]('indent') end,
```
Comment (muniter) [inline runtime/lua/vim/filetype.lua:968]: Is this an issue with the generation?, the wrapping in a function?

File: src/nvim/lua/vim.lua (hunk 1 above)
Comment (clason) [inline src/nvim/lua/vim.lua:666]: I don't think this works as intended... At least CI (and my local `make`) complains _loudly_ about `vim/filetype.lua` not being in the `package.path`...

File: runtime/doc/lua.txt [hunk 3]
```diff
@@ -1778,4 +1778,67 @@ select({items}, {opts}, {on_choice})                         *vim.ui.select()*
...
+                        end
+                        return "bar"
+                      end,
+                    },
+                    filename = {
+                      [".foorc"] = "toml",
+                      ["/etc/foo/config"] = "toml",
+                    },
+                    pattern = {
+                      [".*&zwj;/etc/foo/.*"] = "fooscript",
```
Comment (gpanders) [inline runtime/doc/lua.txt:1823]: I have no idea where this `&zwj;` is coming from 👀

Comment (clason) [reply runtime/doc/lua.txt]: _very_ advanced example

**Comment by gpanders** [comment]
`functionaltest` (but not ASAN or TSAN) is failing with:

```
2022-01-03T20:38:19.8047172Z [  ERROR   ] 1 error, listed below:
2022-01-03T20:38:19.8048125Z [  ERROR   ] test/functional/legacy/memory_usage_spec.lua @ 170: memory usage releases memory when closing windows when folds exist
2022-01-03T20:38:19.8049335Z test/functional/legacy/memory_usage_spec.lua:93: test/helpers.lua:73: Expected objects to be the same.
2022-01-03T20:38:19.8050113Z Passed in:
2022-01-03T20:38:19.8050548Z (boolean) false
2022-01-03T20:38:19.8051041Z Expected:
2022-01-03T20:38:19.8051476Z (boolean) true
```

I don't think it's related, but it's happened twice now. Anyone have any ideas? (tagging @erw7 and @dstein64 since their names show up in `git blame`).

File: runtime/doc/filetype.txt [hunk 4]
```diff
@@ -210,6 +220,34 @@ C. If your file type can be detected by the file name.
       example, for Unix: >
 	:w ~/.config/nvim/filetype.vim
```
Comment (seandewar) [inline runtime/doc/filetype.txt:222]:
```suggestion
<
```

**Comment by neovim-discourse** [comment]
This pull request has been mentioned on **Neovim Discourse**. There might be relevant details there:

https://neovim.discourse.group/t/introducing-filetype-lua-and-a-call-for-help/1806/1

**Comment by tjdevries** [comment]
So we did something similar to this in plenary, but I'm not sure how much applies. I'm just going to post a few links here for some suggestions:

1. We generate the filetype stuff based on github's linguist stuff here: https://github.com/nvim-lua/plenary.nvim/blob/master/scripts/update_filetypes_from_github.lua This provides us with a really large coverage. This generates this file: https://github.com/nvim-lua/plenary.nvim/blob/master/data/plenary/filetypes/base.lua . Just something to consider.
2. It doesn't seem that there is anyway to debug how a particular "rule" is added to filetype, correct? This isn't a huge deal, but it can be annoying trying to track down what rule / plugin / etc. is determing that file `foo` is filetype `bar`.
    - For example, it could be cool if there was a namespace attached to these somehow so you could also remove a namespace if you wanted to (`vim.filetype.remove(vim.api.nvim_create_namespace("vim"))` for example).
3. I think it'd be cool to expose `filetype.match` as something that takes `filetype.match(filepath, contents)` so that you could evaluate what the filetype of arbitrary filepaths and contents would be without loading them into neovim (this is actually one of the main goals of the plenary filetype module, so that we could calculate that ahead of time). Or it can be a different function is fine too. It could even take a function reference for contents so that it doesn't _have_ to load the file contents in if it matches before that.

Overall I love it though :)

**Comment by gpanders** [comment]
Thanks for your input @tjdevries

> 2. It doesn't seem that there is anyway to debug how a particular "rule" is added to filetype, correct? This isn't a huge deal, but it can be annoying trying to track down what rule / plugin / etc. is determing that file foo is filetype bar.
>
>    - For example, it could be cool if there was a namespace attached to these somehow so you could also remove a namespace if you wanted to (vim.filetype.remove(vim.api.nvim_create_namespace("vim")) for example).

I like this idea.

> 3. I think it'd be cool to expose filetype.match as something that takes filetype.match(filepath, contents)

Being able to determine filetype based on content is something I'm actively thinking about. For now we are still punting to `scripts.vim`, but we'll have something along these lines (hopefully) soon.

**Comment by tjdevries** [comment]
:+1: cool.

The other thing is I do not like having mix of generated and post-process content in same tables. I'd prefer to see one the two tables merged via tbl_extend or something so they can be kept completely separate. Or, at least some divider of `-- BEGIN GENERATED` ... `-- END GENERATED`, so that the script can update those in the future if we want (either via tools from linguist, vim patches, etc.)

**Comment by kaiuri** [comment]
I'm a little _short on time_ so I've found it quicker to just add it here (from the link on [filetype.lua -call for help](https://neovim.discourse.group/t/introducing-filetype-lua-and-a-call-for-help/1806)).

The format `*.edn` which is basically just `*.clj`, that is `clojure`, isn't currently recognized by `filetype.lua`. I know, I can merge those tables and be done with it, but `call for help`, right? So, while using the feature I found that Neovim does not recognize `*.edn` files.

---
Unrelated:
Thank you all so much for Neovim, I can't possibly say how much Neovim's been a game changer for me, seriously thank you ALL for saving me from VimScript.

**Comment by clason** [comment]
> The format *.edn which is basically just *.clj, that is clojure, isn't currently recognized by filetype.lua. 

If it's recognized by `filetype.vim`, then this is just one of the patterns that aren't ported yet (which is currently ongoing effort and should be completed soon).

If it's _not_ recognized by `filetype.vim`, then it's an issue for vim/vim, as they are the sole maintainers of the `filetype.vim` runtime file. Any changes made upstream will be ported to Neovim as part of the regular effort to port applicable patches.

**Comment by kaiuri** [comment]
@clason I know, I'm just following up on the post, which asked for `testers`, I actually fixed it myself. Thanks for the follow up.

**Pull request merged**